### DISCUSSION TITLE
`tar_map()` seems to cause dataframes to output as lists

### DISCUSSION THREAD

**Discussion by shirdekel** [ropensci/tarchetypes#76] [Help] [answered]
As the below shows, the output of a dataframe is a dataframe when using `tar_target`, but a list when using `tar_map`. Is this expected behaviour or a bug?

```r
    targets::tar_dir({
      targets::tar_script({
        list(
          targets::tar_target(y, cars)
        )
      })
      targets::tar_make()
      targets::tar_read(y)
    })
    #> • start target y
    #> • built target y
    #> • end pipeline
    #>    speed dist
    #> 1      4    2
    #> 2      4   10
    #> 3      7    4
    #> 4      7   22
    #> 5      8   16
    #> 6      9   10
    #> 7     10   18
    #> 8     10   26
    #> 9     10   34
    #> 10    11   17
    #> 11    11   28
    #> 12    12   14
    #> 13    12   20
    #> 14    12   24
    #> 15    12   28
    #> 16    13   26
    #> 17    13   34
    #> 18    13   34
    #> 19    13   46
    #> 20    14   26
    #> 21    14   36
    #> 22    14   60
    #> 23    14   80
    #> 24    15   20
    #> 25    15   26
    #> 26    15   54
    #> 27    16   32
    #> 28    16   40
    #> 29    17   32
    #> 30    17   40
    #> 31    17   50
    #> 32    18   42
    #> 33    18   56
    #> 34    18   76
    #> 35    18   84
    #> 36    19   36
    #> 37    19   46
    #> 38    19   68
    #> 39    20   32
    #> 40    20   48
    #> 41    20   52
    #> 42    20   56
    #> 43    20   64
    #> 44    22   66
    #> 45    23   54
    #> 46    24   70
    #> 47    24   92
    #> 48    24   93
    #> 49    24  120
    #> 50    25   85

    targets::tar_dir({
      targets::tar_script({
        list(
          tarchetypes::tar_map(
            list(a = list(cars)),
            names = id,
            targets::tar_target(y, a)
          )
        )
      })
      targets::tar_make()
      targets::tar_read(y_355adbf5)
    })
    #> • start target y_355adbf5
    #> • built target y_355adbf5
    #> • end pipeline
    #> [[1]]
    #>  [1]  4  4  7  7  8  9 10 10 10 11 11 12 12 12 12 13 13 13 13 14 14 14 14 15 15
    #> [26] 15 16 16 17 17 17 18 18 18 18 19 19 19 20 20 20 20 20 22 23 24 24 24 24 25
    #> 
    #> [[2]]
    #>  [1]   2  10   4  22  16  10  18  26  34  17  28  14  20  24  28  26  34  34  46
    #> [20]  26  36  60  80  20  26  54  32  40  32  40  50  42  56  76  84  36  46  68
    #> [39]  32  48  52  56  64  66  54  70  92  93 120  85
```
<sup>Created on 2021-09-28 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

<details style="margin-bottom:10px;">
<summary>
Session info
</summary>

    sessioninfo::session_info()
    #> ─ Session info ───────────────────────────────────────────────────────────────
    #>  setting  value                       
    #>  version  R version 4.0.2 (2020-06-22)
    #>  os       macOS Mojave 10.14.6        
    #>  system   x86_64, darwin17.0          
    #>  ui       X11                         
    #>  language (EN)                        
    #>  collate  en_GB.UTF-8                 
    #>  ctype    en_GB.UTF-8                 
    #>  tz       Australia/Sydney            
    #>  date     2021-09-28                  
    #> 
    #> ─ Packages ───────────────────────────────────────────────────────────────────
    #>  package     * version    date       lib source                            
    #>  backports     1.2.1      2020-12-09 [1] CRAN (R 4.0.2)                    
    #>  callr         3.7.0      2021-04-20 [1] CRAN (R 4.0.2)                    
    #>  cli           3.0.1      2021-07-17 [1] CRAN (R 4.0.2)                    
    #>  codetools     0.2-18     2020-11-04 [1] CRAN (R 4.0.2)                    
    #>  crayon        1.4.1      2021-02-08 [1] CRAN (R 4.0.2)                    
    #>  data.table    1.14.2     2021-09-27 [1] CRAN (R 4.0.2)                    
    #>  digest        0.6.28     2021-09-23 [1] CRAN (R 4.0.2)                    
    #>  ellipsis      0.3.2      2021-04-29 [1] CRAN (R 4.0.2)                    
    #>  evaluate      0.14       2019-05-28 [1] CRAN (R 4.0.0)                    
    #>  fansi         0.5.0      2021-05-25 [1] CRAN (R 4.0.2)                    
    #>  fastmap       1.1.0      2021-01-25 [1] CRAN (R 4.0.2)                    
    #>  fs            1.5.0      2020-07-31 [1] CRAN (R 4.0.2)                    
    #>  glue          1.4.2      2020-08-27 [1] CRAN (R 4.0.2)                    
    #>  highr         0.9        2021-04-16 [1] CRAN (R 4.0.2)                    
    #>  htmltools     0.5.2      2021-08-25 [1] CRAN (R 4.0.2)                    
    #>  igraph        1.2.6      2020-10-06 [1] CRAN (R 4.0.2)                    
    #>  knitr         1.34       2021-09-09 [1] CRAN (R 4.0.2)                    
    #>  lifecycle     1.0.1      2021-09-24 [1] CRAN (R 4.0.2)                    
    #>  magrittr      2.0.1      2020-11-17 [1] CRAN (R 4.0.2)                    
    #>  pillar        1.6.3      2021-09-26 [1] CRAN (R 4.0.2)                    
    #>  pkgconfig     2.0.3      2019-09-22 [1] CRAN (R 4.0.0)                    
    #>  processx      3.5.2      2021-04-30 [1] CRAN (R 4.0.2)                    
    #>  ps            1.6.0      2021-02-28 [1] CRAN (R 4.0.2)                    
    #>  purrr         0.3.4      2020-04-17 [1] CRAN (R 4.0.0)                    
    #>  R6            2.5.1      2021-08-19 [1] CRAN (R 4.0.2)                    
    #>  reprex        1.0.0      2021-01-27 [1] CRAN (R 4.0.2)                    
    #>  rlang         0.4.11     2021-04-30 [1] CRAN (R 4.0.2)                    
    #>  rmarkdown     2.11.2     2021-09-25 [1] Github (rstudio/rmarkdown@71ae829)
    #>  sessioninfo   1.1.1      2018-11-05 [1] CRAN (R 4.0.0)                    
    #>  stringi       1.7.4      2021-08-25 [1] CRAN (R 4.0.2)                    
    #>  stringr       1.4.0      2019-02-10 [1] CRAN (R 4.0.0)                    
    #>  styler        1.6.1      2021-09-17 [1] CRAN (R 4.0.2)                    
    #>  targets       0.8.0.9000 2021-09-28 [1] Github (ropensci/targets@65b9557) 
    #>  tibble        3.1.4      2021-08-25 [1] CRAN (R 4.0.2)                    
    #>  tidyselect    1.1.1      2021-04-30 [1] CRAN (R 4.0.2)                    
    #>  utf8          1.2.2      2021-07-24 [1] CRAN (R 4.0.2)                    
    #>  vctrs         0.3.8      2021-04-29 [1] CRAN (R 4.0.2)                    
    #>  withr         2.4.2      2021-04-18 [1] CRAN (R 4.0.2)                    
    #>  xfun          0.26       2021-09-14 [1] CRAN (R 4.0.2)                    
    #>  yaml          2.2.1      2020-02-01 [1] CRAN (R 4.0.0)                    
    #> 
    #> [1] /Library/Frameworks/R.framework/Versions/4.0/Resources/library

</details>

**Comment by wlandau** [comment, marked as answer]
`tar_map()` works by creating the [R expressions](https://adv-r.hadley.nz/expressions.html) for target commands. This works well for atomic scalar values like character strings and numerics (also symbols), but not so well with complicated objects with attributes. In your case, `tar_map()` is trying to turn the `cars` data frame into an R expression that evaluates to that data frame.

``` r
targets::tar_script({
  list(
    tarchetypes::tar_map(
      list(a = list(cars)),
      names = id,
      targets::tar_target(y, a)
    )
  )
})
targets::tar_manifest()$command
#> [1] "list(c(4, 4, 7, 7, 8, 9, 10, 10, 10, 11, 11, 12, 12,  \\n     12, 12, 13, 13, 13, 13, 14, 14, 14, 14, 15, 15, 15, 16, 16,  \\n     17, 17, 17, 18, 18, 18, 18, 19, 19, 19, 20, 20, 20, 20, 20,  \\n     22, 23, 24, 24, 24, 24, 25), c(2, 10, 4, 22, 16, 10, 18,  \\n     26, 34, 17, 28, 14, 20, 24, 28, 26, 34, 34, 46, 26, 36, 60,  \\n     80, 20, 26, 54, 32, 40, 32, 40, 50, 42, 56, 76, 84, 36, 46,  \\n     68, 32, 48, 52, 56, 64, 66, 54, 70, 92, 93, 120, 85))"
```

<sup>Created on 2021-09-28 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Instead, I recommend passing the symbol `cars` instead of the data inside `cars`.

``` r
targets::tar_script({
  list(
    tarchetypes::tar_map(
      list(a = list(as.symbol("cars"))),
      names = id,
      targets::tar_target(y, a)
    )
  )
})
targets::tar_manifest()$command
#> [1] "cars"
```

<sup>Created on 2021-09-28 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Reply by shirdekel** [reply]
Also, I can't get data from packages to work with `tar_make()`:

```r
    palmerpenguins::penguins
    #> # A tibble: 344 × 8
    #>    species island    bill_length_mm bill_depth_mm flipper_length_mm body_mass_g
    #>    <fct>   <fct>              <dbl>         <dbl>             <int>       <int>
    #>  1 Adelie  Torgersen           39.1          18.7               181        3750
    #>  2 Adelie  Torgersen           39.5          17.4               186        3800
    #>  3 Adelie  Torgersen           40.3          18                 195        3250
    #>  4 Adelie  Torgersen           NA            NA                  NA          NA
    #>  5 Adelie  Torgersen           36.7          19.3               193        3450
    #>  6 Adelie  Torgersen           39.3          20.6               190        3650
    #>  7 Adelie  Torgersen           38.9          17.8               181        3625
    #>  8 Adelie  Torgersen           39.2          19.6               195        4675
    #>  9 Adelie  Torgersen           34.1          18.1               193        3475
    #> 10 Adelie  Torgersen           42            20.2               190        4250
    #> # … with 334 more rows, and 2 more variables: sex <fct>, year <int>
    targets::tar_script({
      list(
        tarchetypes::tar_map(
          list(a = list(as.symbol("palmerpenguins::penguins"))),
          names = id,
          targets::tar_target(y, a)
        )
      )
    })
    targets::tar_manifest()
    #> # A tibble: 1 × 3
    #>   name       command                    pattern
    #>   <chr>      <chr>                      <chr>  
    #> 1 y_9d613dba `palmerpenguins::penguins` <NA>
    targets::tar_script({
      list(
        tarchetypes::tar_map(
          list(a = list(as.symbol("palmerpenguins::penguins"))),
          names = id,
          targets::tar_target(y, a)
        )
      )
    })
    targets::tar_make()
    #> • start target y_9d613dba
    #> ✖ error target y_9d613dba
    #> • end pipeline
    #> Error : object 'palmerpenguins::penguins' not found
    #> Error: callr subprocess failed: object 'palmerpenguins::penguins' not found
    #> Visit https://books.ropensci.org/targets/debugging.html for debugging advice.
```
<sup>Created on 2021-09-29 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

<details style="margin-bottom:10px;">
<summary>
Session info
</summary>

    sessioninfo::session_info()
    #> ─ Session info ───────────────────────────────────────────────────────────────
    #>  setting  value                       
    #>  version  R version 4.0.2 (2020-06-22)
    #>  os       macOS Mojave 10.14.6        
    #>  system   x86_64, darwin17.0          
    #>  ui       X11                         
    #>  language (EN)                        
    #>  collate  en_GB.UTF-8                 
    #>  ctype    en_GB.UTF-8                 
    #>  tz       Australia/Sydney            
    #>  date     2021-09-29                  
    #> 
    #> ─ Packages ───────────────────────────────────────────────────────────────────
    #>  package        * version    date       lib source                            
    #>  backports        1.2.1      2020-12-09 [1] CRAN (R 4.0.2)                    
    #>  callr            3.7.0      2021-04-20 [1] CRAN (R 4.0.2)                    
    #>  cli              3.0.1      2021-07-17 [1] CRAN (R 4.0.2)                    
    #>  codetools        0.2-18     2020-11-04 [1] CRAN (R 4.0.2)                    
    #>  crayon           1.4.1      2021-02-08 [1] CRAN (R 4.0.2)                    
    #>  data.table       1.14.2     2021-09-27 [1] CRAN (R 4.0.2)                    
    #>  digest           0.6.28     2021-09-23 [1] CRAN (R 4.0.2)                    
    #>  ellipsis         0.3.2      2021-04-29 [1] CRAN (R 4.0.2)                    
    #>  evaluate         0.14       2019-05-28 [1] CRAN (R 4.0.0)                    
    #>  fansi            0.5.0      2021-05-25 [1] CRAN (R 4.0.2)                    
    #>  fastmap          1.1.0      2021-01-25 [1] CRAN (R 4.0.2)                    
    #>  fs               1.5.0      2020-07-31 [1] CRAN (R 4.0.2)                    
    #>  glue             1.4.2      2020-08-27 [1] CRAN (R 4.0.2)                    
    #>  highr            0.9        2021-04-16 [1] CRAN (R 4.0.2)                    
    #>  htmltools        0.5.2      2021-08-25 [1] CRAN (R 4.0.2)                    
    #>  igraph           1.2.6      2020-10-06 [1] CRAN (R 4.0.2)                    
    #>  knitr            1.34       2021-09-09 [1] CRAN (R 4.0.2)                    
    #>  lifecycle        1.0.1      2021-09-24 [1] CRAN (R 4.0.2)                    
    #>  magrittr         2.0.1      2020-11-17 [1] CRAN (R 4.0.2)                    
    #>  palmerpenguins   0.1.0      2020-07-23 [1] CRAN (R 4.0.2)                    
    #>  pillar           1.6.3      2021-09-26 [1] CRAN (R 4.0.2)                    
    #>  pkgconfig        2.0.3      2019-09-22 [1] CRAN (R 4.0.0)                    
    #>  processx         3.5.2      2021-04-30 [1] CRAN (R 4.0.2)                    
    #>  ps               1.6.0      2021-02-28 [1] CRAN (R 4.0.2)                    
    #>  purrr            0.3.4      2020-04-17 [1] CRAN (R 4.0.0)                    
    #>  R6               2.5.1      2021-08-19 [1] CRAN (R 4.0.2)                    
    #>  reprex           1.0.0      2021-01-27 [1] CRAN (R 4.0.2)                    
    #>  rlang            0.4.11     2021-04-30 [1] CRAN (R 4.0.2)                    
    #>  rmarkdown        2.11.2     2021-09-25 [1] Github (rstudio/rmarkdown@71ae829)
    #>  sessioninfo      1.1.1      2018-11-05 [1] CRAN (R 4.0.0)                    
    #>  stringi          1.7.4      2021-08-25 [1] CRAN (R 4.0.2)                    
    #>  stringr          1.4.0      2019-02-10 [1] CRAN (R 4.0.0)                    
    #>  styler           1.6.1      2021-09-17 [1] CRAN (R 4.0.2)                    
    #>  targets          0.8.0.9000 2021-09-28 [1] Github (ropensci/targets@65b9557) 
    #>  tibble           3.1.4      2021-08-25 [1] CRAN (R 4.0.2)                    
    #>  tidyselect       1.1.1      2021-04-30 [1] CRAN (R 4.0.2)                    
    #>  utf8             1.2.2      2021-07-24 [1] CRAN (R 4.0.2)                    
    #>  vctrs            0.3.8      2021-04-29 [1] CRAN (R 4.0.2)                    
    #>  withr            2.4.2      2021-04-18 [1] CRAN (R 4.0.2)                    
    #>  xfun             0.26       2021-09-14 [1] CRAN (R 4.0.2)                    
    #>  yaml             2.2.1      2020-02-01 [1] CRAN (R 4.0.0)                    
    #> 
    #> [1] /Library/Frameworks/R.framework/Versions/4.0/Resources/library

</details>

**Reply by wlandau** [reply]
`palmerpenguins::penguins` is actually an expression (specifically, a call) with 3 symbols: `::`, `palmerpenguins`, and `penguins`. The actual function call is 

```r
`::`(palmerpenguins, penguins)
```

`quote(palmerpenguins::penguins)` works.

```r
# _targets.R
library(targets)
library(tarchetypes)
tar_map(
  values = list(a = list(quote(palmerpenguins::penguins))), 
  names = id,
  tar_target(y, a)
)
```

**Comment by shirdekel** [comment]
Thanks!

Has this always been the case? I have a project in which I did this and it worked fine.

**Reply by wlandau** [reply]
`tar_map()` did go through some changes, but I do not recall a specific one that would have broken old usage. My recommendation would have been the same regardless, though.
